app/tiebout.hoon

```diff
@@ -0,0 +1,253 @@
+::
+::  +poke-tiebout-action: main action handler
+::
+++  poke-tiebout-action
+  |=  act=action
+  ^-  (quip move _this)
+  ?-  -.act
+    $king
+      (set-king +.act)
+    $token
+      (set-token +.act)
+    $baseurl
+      (set-baseurl +.act)
+    $add-circle
+      (add-circle +.act)
+    $del-circle
+      (del-circle +.act)
+    $notify
+      (send-notify +.act)
+  ==
+::
+::  +add-circle: add circle and subscribe for updates
+::
+++  add-circle
+  |=  nom=name:hall
+  ^-  (quip move _this)
+  :_  this(circles.tie.sta (~(put by circles.tie.sta) nom 0))
+  [ost.bol %peer /our/[nom] [our.bol %hall] /circle/[nom]/config/grams]~
+::
+::  +del-circle: delete circle and unsubscribe from updates
+::
+++  del-circle
+  |=  nom=name:hall
+  ^-  (quip move _this)
+  :_  this(circles.tie.sta (~(del by circles.tie.sta) nom))
+  [ost.bol %pull /our/[nom] [our.bol %hall] ~]~
+::
+::  +set-king: set king @p
+::
+++  set-king
+  |=  kng=@p
+  ^-  (quip move _this)
+  [~ this(king.tie.sta kng)]
+::
+::  +set-token: set iOS device token @t
+::
+++  set-token
+  |=  tok=@t
+  ^-  (quip move _this)
+  [~ this(token.tie.sta tok)]
+::
+::  +set-baseurl: set base url @t
+::
+++  set-baseurl
+  |=  burl=@t
+  ^-  (quip move _this)
+  [~ this(baseurl.tie.sta burl)]
+::
+::  +send-notify: if king, send hiss. if not, do nothing.
+::
+++  send-notify
+  |=  not=notification
+  ^-  (quip move _this)
+  ?:  =(king.tie.sta our.bol)
+    :_  this
+    [ost.bol %hiss /request [~ ~] %httr %hiss (create-apns-request not)]~
+  [~ this]
+::
+::  +diff-hall-prize: receive new circle data
+::
+++  diff-hall-prize
+  |=  [wir=wire piz=prize:hall]
+  ^-  (quip move _this)
+  ~&  prize+[wir piz]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ::
+  ::  %our: set config of circle and iterate through messages, sending
+  ::  notifications for all messages where number is higher than our last-read
+  ::
+      {%our @ @}
+    ?>  ?=(%circle -.piz)
+    =/  nom/name:hall  i.t.wir
+    =/  red/@ud  red.loc.cos.piz
+    [~ this(circles.tie.sta (~(put by circles.tie.sta) nom red))]
+  ==
+::
+::  +reap: recieve acknowledgement for peer
+::
+++  reap
+  |=  [wir=wire err=(unit tang)]
+  ^-  (quip move _this)
+  ::~&  reap+[wir =(~ err)]
+  ?~  err
+    ::  XX send message to users inbox
+    [~ this]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    ~
+  ==
+::
+::  +quit: receive subscription failed, resubscribe
+::
+++  quit
+  |=  wir=wire
+  ^-  (quip move _this)
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    =/  doc/dock  [our.bol %hall]
+    [ost.bol %peer /our/[i.t.wir] doc /circle/[i.t.wir]/config/grams]~
```

Comment:
Maybe we want to re-subscribe from the last-heard message going forward? I think the logic below ensures we don't send notifications for whatever old junk we may get here, but still.

Pretty sure this just gives you the last `100` messages, or all messages from the last `~d1` or something along those lines. Being more specific allows you to catch up from even further back when needed, and doesn't get you more info than you need.